library/src/main/java/org/xmtp/android/library/Conversations.kt

```diff
@@ -433,73 +500,303 @@ data class Conversations(
+                    val envelope = envelopeFromFFi(message)
+                    if (envelope.contentTopic == Topic.userIntro(client.address).description) {
+                        val conversationV1 = fromIntro(envelope = envelope)
+                        if (!streamedConversationTopics.contains(conversationV1.topic)) {
+                            streamedConversationTopics.add(conversationV1.topic)
+                            trySend(conversationV1)
+                        }
+                    }
+
+                    if (envelope.contentTopic == Topic.userInvite(client.address).description) {
+                        val conversationV2 = fromInvite(envelope = envelope)
+                        if (!streamedConversationTopics.contains(conversationV2.topic)) {
+                            streamedConversationTopics.add(conversationV2.topic)
+                            trySend(conversationV2)
+                        }
+                    }
+                }
+            }
+
+            val stream = client.subscribe2(
+                FfiV2SubscribeRequest(
+                    listOf(
+                        Topic.userIntro(client.address).description,
+                        Topic.userInvite(client.address).description
+                    )
+                ),
+                subscriptionCallback
+            )
+
+            awaitClose { launch { stream.end() } }
+        }
+        if (client.v3Client != null) {
+            streamDmConversations()
         }
     }
 
-    fun conversation(sealedInvitation: SealedInvitation): ConversationV2 {
-        val unsealed = sealedInvitation.v1.getInvitation(viewer = client.keys)
-        return ConversationV2.create(
-            client = client,
-            invitation = unsealed,
-            header = sealedInvitation.v1.header,
-        )
+    fun streamAll(): Flow<Conversation> {
+        if (!client.hasV2Client) {
+            return streamConversations()
+        }
+        return merge(streamConversations(), stream())
     }
 
-    /**
+    private fun streamConversations(): Flow<Conversation> = callbackFlow {
+        val conversationCallback = object : FfiConversationCallback {
+            override fun onConversation(conversation: FfiConversation) {
+                if (conversation.groupMetadata().conversationType() == "dm") {
+                    launch {
+                        val dm = Dm(client, conversation)
+                        val matchingConversations = conversationsByTopic.filter {
+                            client.inboxIdFromAddress(it.value.peerAddress)?.lowercase() ==
+                                    dm.peerInboxId().lowercase()
+                        }
+
+                        // Only send the DM if no V2 DM exists
+                        if (matchingConversations.none { it.value.version == Conversation.Version.V2 }) {
+                            trySend(Conversation.Dm(dm)).isSuccess
+                        }
+                    }
```

Comment:
So this is where I think I need the most thinking. Feels like it might be time to keep some sort of map of V2 conversations in libxmtp so we know if the V2 conversation equivalent exists or not. If it exists in V2 then we don't want to send dms for this V3 conversation unless the client does not have a V2 identity. I think the only performant way to list all conversations to do this lookup is going to be if we have a database of the V2 conversations in libxmtp. We don't need all the messages just invites. Thoughts @richardhuaaa?

Reply:
It makes sense to start storing v2 conversations in the libxmtp DB. We eventually need to back up historical v2 conversations anyway once we stop supporting v2. But let's sync up on the other behavior described here!

> If it exists in V2 then we don't want to send dms for this V3 conversation unless the client does not have a V2 identity

I think we would want to send in both, but let's sync up